File: Objective-C/CustomActivityIndicator/CustomActivityIndicator.m
//
//  CustomActivityIndicator.m
//  CustomActivityIndicator
//
//  Created by Dave on 5/25/18.
//  Copyright © 2018 High Tree Development. All rights reserved.
//

#import <UIKit/UIKit.h>
#import "CustomActivityIndicator.h"

@interface CustomActivityIndicator ()

@end

@implementation CustomActivityIndicator

// This class is a singleton. Here we get the shared instance
//
+(CustomActivityIndicator *)shared {
    static CustomActivityIndicator *shared = nil;
    @synchronized(self) {
        if (shared == nil) {
            shared = [[self alloc] init];
        }
    }
    return shared;
}

-(void)show:(UIView *)uiView {
    // Show activity indicator with no animation and no label
    //
    uiView.userInteractionEnabled = NO;
    loadingView = [[UIView alloc] initWithFrame:CGRectMake(0, 0, 80, 80)];
    loadingView.center = uiView.center;
    loadingView.backgroundColor = UIColor.darkGrayColor;
    loadingView.clipsToBounds = YES;
    loadingView.layer.cornerRadius = 10.0;
    
    activityIndicator = [[UIActivityIndicatorView alloc] initWithFrame:CGRectMake(0, 0, 40, 40)];
    activityIndicator.activityIndicatorViewStyle = UIActivityIndicatorViewStyleWhiteLarge;
    activityIndicator.center = CGPointMake(loadingView.frame.size.width / 2, loadingView.frame.size.width / 2);
    activityIndicator.hidesWhenStopped = YES;
    
    dispatch_async(dispatch_get_main_queue(), ^{
        [self->loadingView addSubview:self->activityIndicator];
        [uiView addSubview:self->loadingView];
        [self->activityIndicator startAnimating];
    });
}

-(void)show:(UIView *)uiView backgroundColor:(UIColor *)backgroundColor size:(double)size duration:(double)duration {
    // Show activity indicator with backgroundColor, size, and animates for the duration passsed.
    // If you do not want animation use duration = 0.0
    //
    uiView.userInteractionEnabled = NO;
    loadingView = [[UIView alloc] initWithFrame:CGRectMake(0, 0, size, size)];
    loadingView.center = uiView.center;
    loadingView.backgroundColor = backgroundColor;
    loadingView.clipsToBounds = YES;
    loadingView.layer.cornerRadius = 10.0;
    loadingView.alpha = 0.0;
    
    activityIndicator = [[UIActivityIndicatorView alloc] initWithFrame:CGRectMake(0, 0, size/2, size/2)];
    activityIndicator.activityIndicatorViewStyle = UIActivityIndicatorViewStyleWhiteLarge;
    activityIndicator.center = CGPointMake(loadingView.frame.size.width/2, loadingView.frame.size.width/2);
    activityIndicator.hidesWhenStopped = YES;

    dispatch_async(dispatch_get_main_queue(), ^{
        [self->loadingView addSubview:self->activityIndicator];
        [uiView addSubview:self->loadingView];
        [self->activityIndicator startAnimating];
        
        [UIView animateWithDuration:duration animations:^{
            self->loadingView.alpha = 1.0;
        }];
    });
}

-(void)show:(UIView *)uiView
            backgroundColor:(UIColor *)backgroundColor
            textColor:(UIColor *)textColor
            labelText:(NSString *)labelText
            duration:(double)duration {
    
    // Show activity indicator with background color, text color, label, and animataion
    // Again if you do not want the appearance animated use duration = 0.0
    //
    uiView.userInteractionEnabled = NO;
    label = [[UILabel alloc] initWithFrame:CGRectZero];
    label.font = [UIFont systemFontOfSize:15.0];
    label.numberOfLines = 0;
    [label setText:labelText];
    
    CGSize maxSize = CGSizeMake(200.0, 200.0);
    CGSize requiredSize = [label sizeThatFits:maxSize];
    
    NSLog(@"requiredSize.width: %f", requiredSize.width);
    NSLog(@"requiredSize.height: %f", requiredSize.height);
    
    if (requiredSize.width > maxSize.width || requiredSize.height > maxSize.height) {
        // Label is too big
        //
        [self show:uiView backgroundColor:backgroundColor size:80.0 duration:duration];
    } else {
        CGFloat height = requiredSize.height + 100;
        
        NSLog(@"height: %f", height);
        NSLog(@"requiredSize.height: %f", requiredSize.height);
        
        loadingView = [[UIView alloc]initWithFrame:CGRectMake(0, 0, requiredSize.width + 30, height)];
        loadingView.center = uiView.center;
        label = [[UILabel alloc]initWithFrame:CGRectMake(0, 0, requiredSize.width + 20, requiredSize.height + 18)];
        label.lineBreakMode = NSLineBreakByWordWrapping;
        label.numberOfLines = 0;
        label.center = CGPointMake(loadingView.frame.size.width / 2, loadingView.frame.size.height - requiredSize.height);
        activityIndicator = [[UIActivityIndicatorView alloc] initWithFrame:CGRectMake(0, 0, 40, 40)];
        activityIndicator.center = CGPointMake(loadingView.frame.size.width/2, 50.0);
    }
    
    label.text = labelText;
    [label setTextAlignment:NSTextAlignmentCenter];
    label.textColor = UIColor.whiteColor;
    
    loadingView.backgroundColor = backgroundColor;
    loadingView.clipsToBounds = YES;
    loadingView.layer.cornerRadius = 10.0;
    loadingView.alpha = 0.0;
    
    activityIndicator.activityIndicatorViewStyle = UIActivityIndicatorViewStyleWhiteLarge;
    activityIndicator.hidesWhenStopped = YES;
    
    dispatch_async(dispatch_get_main_queue(), ^{
        NSLog(@"label.width: %f", self->label.frame.size.width);
        NSLog(@"label.height: %f", self->label.frame.size.height);
        [self->loadingView addSubview:self->label];
        [self->loadingView addSubview:self->activityIndicator];
        [uiView addSubview:self->loadingView];
        [self->activityIndicator startAnimating];
        
        [UIView animateWithDuration:duration animations:^{
            self->loadingView.alpha = 1.0;
        }];
    });
}

-(void)hide:(UIView *)uiView {
    if ([loadingView isDescendantOfView:uiView]) {
        dispatch_async(dispatch_get_main_queue(), ^{
            uiView.userInteractionEnabled = YES;
            [self->activityIndicator stopAnimating];
            [self->loadingView removeFromSuperview];
        });
    } else {
        // add a short delay incase hide was called before the view appeared
        //
        dispatch_time_t popTime = dispatch_time(DISPATCH_TIME_NOW, (int64_t)(0.2 * NSEC_PER_SEC));
        dispatch_after(popTime, dispatch_get_main_queue(), ^(void){
            if ([self->loadingView isDescendantOfView:uiView]) {
                uiView.userInteractionEnabled = YES;
                [self->activityIndicator stopAnimating];
                [self->loadingView removeFromSuperview];
            }
        });
    }
}

-(void)hide:(UIView *)uiView duration:(double)duration {
    if ([loadingView isDescendantOfView:uiView]) {
        dispatch_async(dispatch_get_main_queue(), ^{
            [UIView animateWithDuration:duration animations:^{
                self->loadingView.alpha = 0.0;
            } completion:^(BOOL finished) {
                if (finished) {
                    uiView.userInteractionEnabled = YES;
                    [self->activityIndicator stopAnimating];
                    [self->loadingView removeFromSuperview];
                }
            }];
        });
    } else {
        dispatch_time_t popTime = dispatch_time(DISPATCH_TIME_NOW, (int64_t)(0.2 * NSEC_PER_SEC));
        dispatch_after(popTime, dispatch_get_main_queue(), ^(void){
            if ([self->loadingView isDescendantOfView:uiView]) {
                [UIView animateWithDuration:duration animations:^{
                    self->loadingView.alpha = 0.0;
                } completion:^(BOOL finished) {
                    if (finished) {
                        uiView.userInteractionEnabled = YES;
                        [self->activityIndicator stopAnimating];
                        [self->loadingView removeFromSuperview];
                    }
                }];
            }
        });
    }
}

@end

File: Swift/CustomActivityIndicator/CustomActivityIndicator.swift
//
//  CustomActivityIndicator.swift
//  AnatomyShare
//
//  Created by Dave on 1/12/18.
//  Copyright © 2018 Dave. All rights reserved.
//

import UIKit

final class CustomActivityIndicator {
    
    // We need to make this a singleton in order for
    // the same views to be added and removed each time
    //
    static let shared = CustomActivityIndicator()
    
    // Init is private so another instance of this class cannot be created
    //
    private init() {}
    
    var loadingView: UIView = UIView()
    var activityIndicator: UIActivityIndicatorView = UIActivityIndicatorView()
    var label: UILabel = UILabel()
    
    // Show customized activity indicator,
    // actually add activity indicator to passing view
    // @param uiView - add activity indicator to this view
    // @param backgroundColor - color for the background of the activity indicator
    // @param size - desired size of the activity indicator
    // @parm animated - Bool animate the appearance (fade in / out)
    // @parm duration - Length of time for the animation
    //
    func showActivityIndicator(uiView: UIView,
                               backgroundColor: UIColor = .darkGray,
                               size: Double = 80,
                               animated: Bool = false,
                               duration: Double = 1.0) {
        
        uiView.isUserInteractionEnabled = false
        loadingView.frame = CGRect(x: 0, y: 0, width: size, height: size)
        loadingView.center = uiView.center
        loadingView.backgroundColor = backgroundColor
        loadingView.clipsToBounds = true
        loadingView.layer.cornerRadius = 10
        
        if animated {
            // Set alpha to allow fading in / out
            loadingView.alpha = 0.0
        } else {
            loadingView.alpha = 1.0
        }
        
        activityIndicator.frame = CGRect(x: 0, y: 0, width: size / 2, height: size / 2)
        activityIndicator.activityIndicatorViewStyle =
            UIActivityIndicatorViewStyle.whiteLarge
        activityIndicator.center = CGPoint(x: loadingView.frame.size.width / 2, y: loadingView.frame.size.height / 2)
        activityIndicator.hidesWhenStopped = true
        
        DispatchQueue.main.async {
            self.loadingView.addSubview(self.activityIndicator)
            uiView.addSubview(self.loadingView)
            self.activityIndicator.startAnimating()
            if animated {
                // Animate the appearance
                self.loadingView.fadeIn(duration: duration)
            }
        }
    }
    
    // Show customized activity indicator,
    // actually add activity indicator to passing view
    // @param uiView - add activity indicator to this view
    // @parm labelText - text to add to activity indicator
    // @param backgroundColor - color for the background of the activity indicator
    // @param textColor - desired color the the label text
    // @parm animated - Bool animate the appearance (fade in / out)
    // @parm duration - Length of time for the animation
    //
    func showActivityIndicator(uiView: UIView,
                               labelText: String,
                               backgroundColor: UIColor = .darkGray,
                               textColor: UIColor = .white,
                               animated: Bool = false,
                               duration: Double = 1.0) {
        
        uiView.isUserInteractionEnabled = false
        let width: CGFloat = labelText.width(withConstrainedHeight: 21.0, font: UIFont.systemFont(ofSize: 17))
        
        // Max width is 200 if width is greater than this we need to create a multi-lined label
        //
        if width > 200 {
            let height: CGFloat = labelText.height(withConstrainedWidth: 200, font: UIFont.systemFont(ofSize: 17))
            if height > 200 {
                // Too Large just show a regular indicator
                //
                showActivityIndicator(uiView: uiView, backgroundColor: backgroundColor)
                return
            }
            loadingView.frame = CGRect(x: 0, y: 0, width: 220, height: 110 + height)
            loadingView.center = uiView.center
            label = UILabel(frame: CGRect(x: 0, y: 0, width: 200, height: height))
            label.center = CGPoint(x: loadingView.frame.size.width / 2, y: (height / 2) + 80)
            label.numberOfLines = 0
            activityIndicator.frame = CGRect(x: 0, y: 0, width: 40, height: 40)
            activityIndicator.center = CGPoint(x: 110, y: 50)
            
        } else {
            loadingView.frame = CGRect(x: 0, y: 0, width: width + 20, height: 110)
            loadingView.center = uiView.center
            label = UILabel(frame: CGRect(x: 0, y: 0, width: width, height: 21))
            label.center = CGPoint(x: loadingView.frame.size.width / 2, y: 90)
            activityIndicator.frame = CGRect(x: 0, y: 0, width: 40, height: 40)
            activityIndicator.center = CGPoint(x: loadingView.frame.size.width / 2, y: loadingView.frame.size.height / 2)
        }
        
        loadingView.backgroundColor = backgroundColor
        loadingView.clipsToBounds = true
        loadingView.layer.cornerRadius = 10
        
        label.textAlignment = .center
        label.textColor = UIColor.white
        label.text = labelText
        
        activityIndicator.activityIndicatorViewStyle = UIActivityIndicatorViewStyle.whiteLarge
        activityIndicator.hidesWhenStopped = true
        
        if animated {
            // Set alpha to allow fading in / out
            loadingView.alpha = 0.0
        } else {
            loadingView.alpha = 1.0
        }
        
        DispatchQueue.main.async {
            self.loadingView.addSubview(self.label)
            self.loadingView.addSubview(self.activityIndicator)
            uiView.addSubview(self.loadingView)
            self.activityIndicator.startAnimating()
            if animated {
                // Animate the appearance
                self.loadingView.fadeIn(duration: duration)
            }
        }
    }
    
    // Hide activity indicator
    // Actually remove activity indicator from its super view
    // @param uiView - remove activity indicator from this view
    // @parm animated - Bool animate the appearance (fade in / out)
    // @parm duration - Length of time for the animation
    //
    func hideActivityIndicator(uiView: UIView, animated: Bool = false, duration: Double = 1.0) {
        // check to make sure container is a subview before we
        // remove it
        if loadingView.isDescendant(of: uiView) {
            DispatchQueue.main.async {
                if animated {
                    // Fade the activity indicator out
                    UIView.animate(withDuration: duration, animations: {
                        self.loadingView.alpha = 0.0
                    }, completion: { finished in
                        if finished {
                            uiView.isUserInteractionEnabled = true
                            self.activityIndicator.stopAnimating()
                            self.loadingView.removeFromSuperview()
                            self.label.text = ""
                        }
                    })
                } else {
                    uiView.isUserInteractionEnabled = true
                    self.activityIndicator.stopAnimating()
                    self.loadingView.removeFromSuperview()
                    self.label.text = ""
                }
            }
        } else {
            // Wait a little bit and check again. Hide could have been called before the original indicator appeared
            //
            DispatchQueue.main.asyncAfter(deadline: .now() + 0.2) {
                if self.loadingView.isDescendant(of: uiView) {
                    if animated {
                        // Fade the activity indicator out
                        UIView.animate(withDuration: duration, animations: {
                            self.loadingView.alpha = 0.0
                        }, completion: { finished in
                            if finished {
                                uiView.isUserInteractionEnabled = true
                                self.activityIndicator.stopAnimating()
                                self.loadingView.removeFromSuperview()
                                self.label.text = ""
                            }
                        })
                    } else {
                        uiView.isUserInteractionEnabled = true
                        self.activityIndicator.stopAnimating()
                        self.loadingView.removeFromSuperview()
                        self.label.text = ""
                    }
                }
            }
        }
    }
    
    // Hide activity indicator
    // Actually remove activity indicator from its super view
    // @param uiView - remove activity indicator from this view
    // @param delay - milliseconds to delay the removal of the activity indicator
    // @parm animated - Bool animate the appearance (fade in / out)
    // @parm duration - Length of time for the animation
    //
    func hideActivityIndicator(uiView: UIView, delay: Double, animated: Bool = false, duration: Double = 1.0) {
        DispatchQueue.main.asyncAfter(deadline: .now() + delay) {
            if self.loadingView.isDescendant(of: uiView) {
                if animated {
                    // Fade the activity indicator out
                    UIView.animate(withDuration: duration, animations: {
                        self.loadingView.alpha = 0.0
                    }, completion: { finished in
                        if finished {
                            uiView.isUserInteractionEnabled = true
                            self.activityIndicator.stopAnimating()
                            self.loadingView.removeFromSuperview()
                            self.label.text = ""
                        }
                    })
                } else {
                    uiView.isUserInteractionEnabled = true
                    self.activityIndicator.stopAnimating()
                    self.loadingView.removeFromSuperview()
                    self.label.text = ""
                }
            }
        }
    }
}

// Sample calls to show indicator
//

// CustomActivityIndicator.shared.showActivityIndicator(uiView: self.view)
// CustomActivityIndicator.shared.showActivityIndicator(uiView: self.view, animated: true)
// CustomActivityIndicator.shared.showActivityIndicator(uiView: self.view, animated: true, duration: 0.5)
// CustomActivityIndicator.shared.showActivityIndicator(uiView: self.view, backgroundColor: .red)
// CustomActivityIndicator.shared.showActivityIndicator(uiView: self.view, size: 200)
// CustomActivityIndicator.shared.showActivityIndicator(uiView: self.view, backgroundColor: .black, size: 100)
//
// CustomActivityIndicator.shared.showActivityIndicator(uiView: self.view, labelText: "Your Text Here")
// CustomActivityIndicator.shared.showActivityIndicator(uiView: self.view, labelText: "Your Text Here", backgroundColor: .red)
// CustomActivityIndicator.shared.showActivityIndicator(uiView: self.view, labelText: "Your Text Here", textColor: .red)
// CustomActivityIndicator.shared.showActivityIndicator(uiView: self.view, labelText: "Your Text Here", backgroundColor: .black, textColor: .red)

// CustomActivityIndicator.shared.hideActivityIndicator(uiView: self.view)
// CustomActivityIndicator.shared.hideActivityIndicator(uiView: self.view, animated: true)
// CustomActivityIndicator.shared.hideActivityIndicator(uiView: self.view, animated: true, duration: 2.0)

extension UIView {
    // Fade in a view with a duration
    func fadeIn(duration: TimeInterval = 1.0) {
        UIView.animate(withDuration: duration, animations: {
            self.alpha = 1.0
        })
    }
    
    // Fade out a view with a duration (Not Used)
    func fadeOut(duration: TimeInterval = 1.0) {
        UIView.animate(withDuration: duration, animations: {
            self.alpha = 0.0
        })
    }
}

// Required extensions to calculate the size of the activity indicator when text for a label is used
//
extension String {
    
    func height(withConstrainedWidth width: CGFloat, font: UIFont) -> CGFloat {
        let constraintRect = CGSize(width: width, height: .greatestFiniteMagnitude)
        let boundingBox = self.boundingRect(with: constraintRect, options: .usesLineFragmentOrigin, attributes: [NSAttributedStringKey.font: font], context: nil)
        
        return ceil(boundingBox.height)
    }
    
    func width(withConstrainedHeight height: CGFloat, font: UIFont) -> CGFloat {
        let constraintRect = CGSize(width: .greatestFiniteMagnitude, height: height)
        let boundingBox = self.boundingRect(with: constraintRect, options: .usesLineFragmentOrigin, attributes: [NSAttributedStringKey.font: font], context: nil)
        
        return ceil(boundingBox.width)
    }
}
